### PR TITLE
Fix regression in DataFrame.write_xxx

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-ballista/issues/894

Status

- [x] Add failing test
- [x] Upgrade to DataFusion version that supports CopyTo serde (https://github.com/apache/arrow-datafusion/pull/8618)
- [x] Upgrade to DataFusion version that supports CopyTo serde with support for WriterProperties (https://github.com/apache/arrow-datafusion/pull/8627 and https://github.com/apache/arrow-datafusion/pull/8641)
- [x] Upgrade to DataFusion version that supports serde for Csv and Parquet FileSinkExec (https://github.com/apache/arrow-datafusion/pull/8646)
- [x] Test passes
- [x] Upgrade to DataFusion main branch version
- [x] Clippy fixes for latest Rust version

 # Rationale for this change

Fix a major regression where we can no longer save the results of a query. 

# What changes are included in this PR?

- Upgrade to more recent DataFusion
- Add test

# Are there any user-facing changes?

### CODE REVIEW

File: ballista/cache/src/listener/loading_cache.rs [hunk 1]
```diff
@@ -145,7 +145,7 @@ where
     fn listen_on_get_if_present(&self, k: Self::K, v: Option<Self::V>) {
         if self.listeners.len() == 1 {
             self.listeners
-                .get(0)
+                .first()
```
Comment (andygrove) [inline ballista/cache/src/listener/loading_cache.rs:148]: Clippy

**Review by Dandandan — APPROVED** [review]

File: ballista/client/src/context.rs [hunk 2]
```diff
@@ -462,12 +462,15 @@ impl BallistaContext {
 }
 
 #[cfg(test)]
-mod tests {
-    #[cfg(feature = "standalone")]
+#[cfg(feature = "standalone")]
+mod standalone_tests {
```
Comment (andygrove) [inline ballista/client/src/context.rs:466]: Drive-by cleanup to declare that the whole module requires the `standalone` feature rather than declare this on each import and each test.
